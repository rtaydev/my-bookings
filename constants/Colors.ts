/**
 * Below are the colors that are used in the app. The colors are defined in the light and dark mode.
 * There are many other ways to style your app. For example, [Nativewind](https://www.nativewind.dev/), [Tamagui](https://tamagui.dev/), [unistyles](https://reactnativeunistyles.vercel.app), etc.
 */

const tintColorLight = "#000";
const tintColorDark = "#fff";

export const Colors = {
  light: {
    text: "#000",
    background: "#fff",
    secondaryBackground: "#0030831A",
    tint: tintColorLight,
    icon: "#003083",
    tabIconDefault: "#4dd0e1",
    tabIconSelected: tintColorLight,
    buttonBackground: "#003083",
    buttonText: "#fff",
    border: "#003083",
    headingTextColor: "#003083",
  },
  dark: {
    text: "#fff",
    background: "#003083",
    secondaryBackground: "#fff",
    tint: tintColorDark,
    icon: "#000",
    tabIconDefault: "#26a69a",
    tabIconSelected: tintColorDark,
    buttonBackground: "#fff",
    buttonText: "#003083",
    border: "#efefef",
    headingTextColor: "#fff",
  },
};
