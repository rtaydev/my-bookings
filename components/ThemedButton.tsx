import React from "react";
import {
  StyleSheet,
  Text,
  Pressable,
  Linking,
  Platform,
  View,
} from "react-native";
import { useThemeColor } from "@/hooks/useThemeColor";
import { useColorScheme } from "react-native";
import { Ionicons } from "@expo/vector-icons";
import { Colors } from "@/constants/Colors";

interface ThemedButtonProps {
  title?: string;
  lightColor?: string;
  darkColor?: string;
  style?: object;
  onPress?: () => void;
  testID?: string;
  href?: string;
  icon?: string;
  wFull?: boolean;
  variant?: "default" | "outline" | "text";
}

export const ThemedButton: React.FC<ThemedButtonProps> = ({
  title,
  lightColor,
  darkColor,
  style,
  onPress,
  testID,
  href,
  icon,
  wFull = false,
  variant = "default",
}) => {
  const backgroundColor = useThemeColor(
    { light: lightColor, dark: darkColor },
    "buttonBackground",
  );

  const handlePress = () => {
    if (href && Platform.OS === "web") {
      Linking.openURL(href);
    } else if (onPress) {
      onPress();
    }
  };

  const colorScheme = useColorScheme();
  const textColor =
    variant === "default"
      ? Colors[colorScheme ?? "light"].buttonText
      : backgroundColor;

  const getVariantStyles = () => {
    switch (variant) {
      case "outline":
        return {
          backgroundColor: "transparent",
          borderWidth: 1,
          borderColor: backgroundColor,
        };
      case "text":
        return {
          backgroundColor: "transparent",
        };
      default:
        return {
          backgroundColor,
        };
    }
  };

  return (
    <Pressable
      onPress={handlePress}
      style={({ pressed }) => [
        styles.button,
        getVariantStyles(),
        { opacity: pressed ? 0.8 : 1 },
        style,
        !title && icon && styles.iconOnlyButton,
        wFull ? styles.fullWidthButton : styles.autoWidthButton, // Adjust width based on wFull
      ]}
      testID={testID}
      accessibilityRole="button"
    >
      <View style={styles.content}>
        {icon && (
          <Ionicons
            name={icon}
            style={[styles.icon, { color: textColor }]}
            size={16}
          />
        )}
        {title && (
          <Text
            style={[
              styles.buttonText,
              { color: textColor },
              icon && { marginLeft: 8 },
            ]}
          >
            {title}
          </Text>
        )}
      </View>
    </Pressable>
  );
};

const styles = StyleSheet.create({
  button: {
    padding: 10,
    borderRadius: 30,
    alignItems: "center",
    justifyContent: "center",
  },
  buttonText: {
    fontSize: 16,
  },
  content: {
    flexDirection: "row",
    alignItems: "center",
  },
  iconOnlyButton: {
    width: 40,
    height: 40,
    borderRadius: 20,
  },
  fullWidthButton: {
    width: "100%",
  },
  autoWidthButton: {
    alignSelf: "flex-start",
  },
});
