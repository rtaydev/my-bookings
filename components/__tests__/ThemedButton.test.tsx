import React from "react";
import { render, fireEvent } from "@testing-library/react-native";
import { ThemedButton } from "../ThemedButton";
import { Colors } from "@/constants/Colors";

describe("ThemedButton", () => {
  it("renders correctly with theme", () => {
    const { getByTestId, toJSON } = render(
      <ThemedButton
        title="Test Button"
        darkColor="#1E88E5"
        testID="themed-button"
      />,
    );

    const button = getByTestId("themed-button");
    const flattenedStyle = button.props.style
      .flat()
      .reduce((acc, style) => ({ ...acc, ...style }), {});

    expect(flattenedStyle).toEqual(
      expect.objectContaining({
        backgroundColor: Colors.light.buttonBackground,
      }),
    );
    expect(toJSON()).toMatchSnapshot();
  });

  it("applies custom styles", () => {
    const { getByTestId, toJSON } = render(
      <ThemedButton
        title="Test Button"
        style={{ padding: 20 }}
        testID="themed-button"
      />,
    );

    const button = getByTestId("themed-button");
    const flattenedStyle = button.props.style
      .flat()
      .reduce((acc, style) => ({ ...acc, ...style }), {});

    expect(flattenedStyle).toEqual(expect.objectContaining({ padding: 20 }));
    expect(toJSON()).toMatchSnapshot();
  });

  it("passes other props to Pressable", () => {
    const onPressMock = jest.fn();
    const { getByTestId, toJSON } = render(
      <ThemedButton
        title="Test Button"
        onPress={onPressMock}
        testID="themed-button"
      />,
    );

    const button = getByTestId("themed-button");
    fireEvent.press(button);
    expect(onPressMock).toHaveBeenCalled();
    expect(toJSON()).toMatchSnapshot();
  });

  it("matches the snapshot", () => {
    const { toJSON } = render(
      <ThemedButton
        title="Snapshot Button"
        darkColor="#1E88E5"
        testID="snapshot-button"
      />,
    );
    expect(toJSON()).toMatchSnapshot();
  });
});
