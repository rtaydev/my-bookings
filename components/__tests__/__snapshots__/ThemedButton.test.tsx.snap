// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ThemedButton applies custom styles 1`] = `
<View
  accessibilityRole="button"
  accessibilityState={
    {
      "busy": undefined,
      "checked": undefined,
      "disabled": undefined,
      "expanded": undefined,
      "selected": undefined,
    }
  }
  accessibilityValue={
    {
      "max": undefined,
      "min": undefined,
      "now": undefined,
      "text": undefined,
    }
  }
  accessible={true}
  collapsable={false}
  focusable={true}
  onBlur={[Function]}
  onClick={[Function]}
  onFocus={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    [
      {
        "alignItems": "center",
        "borderRadius": 30,
        "justifyContent": "center",
        "padding": 10,
      },
      {
        "backgroundColor": "#003083",
      },
      {
        "opacity": 1,
      },
      {
        "padding": 20,
      },
      false,
      {
        "alignSelf": "flex-start",
      },
    ]
  }
  testID="themed-button"
>
  <View
    style={
      {
        "alignItems": "center",
        "flexDirection": "row",
      }
    }
  >
    <Text
      style={
        [
          {
            "fontSize": 16,
          },
          {
            "color": "#fff",
          },
          undefined,
        ]
      }
    >
      Test Button
    </Text>
  </View>
</View>
`;

exports[`ThemedButton matches the snapshot 1`] = `
<View
  accessibilityRole="button"
  accessibilityState={
    {
      "busy": undefined,
      "checked": undefined,
      "disabled": undefined,
      "expanded": undefined,
      "selected": undefined,
    }
  }
  accessibilityValue={
    {
      "max": undefined,
      "min": undefined,
      "now": undefined,
      "text": undefined,
    }
  }
  accessible={true}
  collapsable={false}
  focusable={true}
  onBlur={[Function]}
  onClick={[Function]}
  onFocus={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    [
      {
        "alignItems": "center",
        "borderRadius": 30,
        "justifyContent": "center",
        "padding": 10,
      },
      {
        "backgroundColor": "#003083",
      },
      {
        "opacity": 1,
      },
      undefined,
      false,
      {
        "alignSelf": "flex-start",
      },
    ]
  }
  testID="snapshot-button"
>
  <View
    style={
      {
        "alignItems": "center",
        "flexDirection": "row",
      }
    }
  >
    <Text
      style={
        [
          {
            "fontSize": 16,
          },
          {
            "color": "#fff",
          },
          undefined,
        ]
      }
    >
      Snapshot Button
    </Text>
  </View>
</View>
`;

exports[`ThemedButton passes other props to Pressable 1`] = `
<View
  accessibilityRole="button"
  accessibilityState={
    {
      "busy": undefined,
      "checked": undefined,
      "disabled": undefined,
      "expanded": undefined,
      "selected": undefined,
    }
  }
  accessibilityValue={
    {
      "max": undefined,
      "min": undefined,
      "now": undefined,
      "text": undefined,
    }
  }
  accessible={true}
  collapsable={false}
  focusable={true}
  onBlur={[Function]}
  onClick={[Function]}
  onFocus={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    [
      {
        "alignItems": "center",
        "borderRadius": 30,
        "justifyContent": "center",
        "padding": 10,
      },
      {
        "backgroundColor": "#003083",
      },
      {
        "opacity": 1,
      },
      undefined,
      false,
      {
        "alignSelf": "flex-start",
      },
    ]
  }
  testID="themed-button"
>
  <View
    style={
      {
        "alignItems": "center",
        "flexDirection": "row",
      }
    }
  >
    <Text
      style={
        [
          {
            "fontSize": 16,
          },
          {
            "color": "#fff",
          },
          undefined,
        ]
      }
    >
      Test Button
    </Text>
  </View>
</View>
`;

exports[`ThemedButton renders correctly with theme 1`] = `
<View
  accessibilityRole="button"
  accessibilityState={
    {
      "busy": undefined,
      "checked": undefined,
      "disabled": undefined,
      "expanded": undefined,
      "selected": undefined,
    }
  }
  accessibilityValue={
    {
      "max": undefined,
      "min": undefined,
      "now": undefined,
      "text": undefined,
    }
  }
  accessible={true}
  collapsable={false}
  focusable={true}
  onBlur={[Function]}
  onClick={[Function]}
  onFocus={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    [
      {
        "alignItems": "center",
        "borderRadius": 30,
        "justifyContent": "center",
        "padding": 10,
      },
      {
        "backgroundColor": "#003083",
      },
      {
        "opacity": 1,
      },
      undefined,
      false,
      {
        "alignSelf": "flex-start",
      },
    ]
  }
  testID="themed-button"
>
  <View
    style={
      {
        "alignItems": "center",
        "flexDirection": "row",
      }
    }
  >
    <Text
      style={
        [
          {
            "fontSize": 16,
          },
          {
            "color": "#fff",
          },
          undefined,
        ]
      }
    >
      Test Button
    </Text>
  </View>
</View>
`;
